samcli/lib/utils/async_utils.py

```diff
@@ -0,0 +1,131 @@
+    task is thrown an exception during its execution
+    """
+    async_tasks = []
+    results = []
+
+    LOG.debug("Async execution started")
+
+    for task in tasks:
+        # loop in all tasks and pass them to the executor
+        LOG.debug("Invoking function %s", str(task.function))
+        async_tasks.append(event_loop.run_in_executor(None, task.function, *task.args))
+
+    LOG.debug("Waiting for async results")
+
+    for result in await asyncio.gather(*async_tasks, return_exceptions=True):
+        # for each task, wait for them to complete
+        if isinstance(result, Exception):
+            LOG.debug("Exception raised during the execution")
+            # if the result is a type of Exception, stop the event loop and raise it back to caller
+            raise result
+        results.append(result)
+
+    LOG.debug("Async execution completed")
+
+    # flush all loggers which is printed during async execution
+    if logging.root and logging.root.handlers:
+        for handler in logging.root.handlers:
+            handler.flush()
+
+    return results
+
+
+def run_given_tasks_async(tasks, event_loop=asyncio.get_event_loop()):
+    """
+    Runs the given list of tasks in the given (or default) event loop.
+    This function will wait for execution to be completed
+
+    See _run_given_tasks_async for details
+
+    Parameters
+    ----------
+    tasks: list of Task definitions that will be executed
+
+    event_loop: EventLoop instance that will be used to execute these tasks
+
+    Returns
+    -------
+    List of results from the given Task list. Raises the exception if any of the underlying functions throw one
+    """
+    return event_loop.run_until_complete(_run_given_tasks_async(tasks, event_loop))
+
+
+class AsyncContext:
+    """
+    A helper class to hold list of tasks, and manages their execution
+    """
+
+    def __init__(self):
+        self._async_tasks = []
+
+    def add_async_task(self, function, args=tuple()):
+        """
+        Add a function definition and its args to the the async context, which will be executed later
+
+        Parameters
+        ----------
+        function: Reference to the function definition which will be executed in this AsyncContext
+
+        args: Parameters of the function which will be executed
+        """
+        self._async_tasks.append(Task(function, args))
+
+    def run_async(self):
+        """
+        Will run all collected functions in async context, and return their results in order
+
+        Returns
+        -------
+        List of result of the executions in order
+        """
+        event_loop = asyncio.new_event_loop()
+        return run_given_tasks_async(self._async_tasks, event_loop)
+
+
+class Task:
+    """
+    This class is used to define a task which consists of a function definition and the parameters
+    """
+
+    def __init__(self, function, args=tuple()):
+        self.function = function
+        self.args = args
```

Comment:
We can use [`functool.partial()`](https://docs.python.org/3.9/library/functools.html#functools.partial) instead, e.g.

```python
task = partial(function, args)
task()
```

Allows getting rid of this class, plays well with `asyncio` (e.g. for passing keyword arguments, which isn't possible without `partial`) and simplifies `AsyncContext` (can directly append the partial function).